src/main/java/org/broadinstitute/hellbender/utils/smithwaterman/SmithWatermanJavaAligner.java

```diff
@@ -65,33 +70,95 @@ public SmithWatermanAlignment align(final byte[] reference, final byte[] alterna
         Utils.nonNull(overhangStrategy);
 
         // avoid running full Smith-Waterman if there is an exact match of alternate in reference
-        int matchIndex = -1;
+        int exactMatchIndex = -1;
         if (overhangStrategy == SWOverhangStrategy.SOFTCLIP || overhangStrategy == SWOverhangStrategy.IGNORE) {
             // Use a substring search to find an exact match of the alternate in the reference
             // NOTE: This approach only works for SOFTCLIP and IGNORE overhang strategies
-            matchIndex = Utils.lastIndexOf(reference, alternate);
+            exactMatchIndex = Utils.lastIndexOf(reference, alternate);
         }
 
         final SmithWatermanAlignment alignmentResult;
 
-        if (matchIndex != -1) {
+        //if exact match
+        if (exactMatchIndex != -1) {
             // generate the alignment result when the substring search was successful
-            final List<CigarElement> lce = new ArrayList<>(alternate.length);
-            lce.add(makeElement(State.MATCH, alternate.length));
-            alignmentResult = new SWPairwiseAlignmentResult(AlignmentUtils.consolidateCigar(new Cigar(lce)), matchIndex);
+            final List<CigarElement> lce = Collections.singletonList(makeElement(State.MATCH, alternate.length));
+            alignmentResult = new SWPairwiseAlignmentResult(AlignmentUtils.consolidateCigar(new Cigar(lce)), exactMatchIndex);
         }
         else {
-            // run full Smith-Waterman
-            final int n = reference.length+1;
-            final int m = alternate.length+1;
-            final int[][] sw = new int[n][m];
-            final int[][] btrack=new int[n][m];
-
-            calculateMatrix(reference, alternate, sw, btrack, overhangStrategy, parameters);
-            alignmentResult = calculateCigar(sw, btrack, overhangStrategy); // length of the segment (continuous matches, insertions or deletions)
+            int singleMismatchIndex = -1;
+            if (overhangStrategy == SWOverhangStrategy.SOFTCLIP || overhangStrategy == SWOverhangStrategy.IGNORE) {
+                //look for one mismatch
+                singleMismatchIndex = Utils.lastIndexOfAtMostOneMismatch(reference, alternate);
+            }
+
+            if (singleMismatchIndex != -1) {
+                // generate the alignment result when the substring search was successful
+                final List<CigarElement> lce = Collections.singletonList(makeElement(State.MATCH, alternate.length));
+                alignmentResult = new SWPairwiseAlignmentResult(AlignmentUtils.consolidateCigar(new Cigar(lce)), singleMismatchIndex);
+            }
+            else{
+                if(haplotypeToref){
```

Comment:
Can you extract a private method somewhere around here?  It's getting pretty dense.